binary-transcoders/src/main/kotlin/com/kamelia/sprinkler/transcoder/binary/decoder/composer/DecodingScopeImpl.kt

```diff
@@ -0,0 +1,156 @@
+
+    @JvmName("decodeByte")
+    override fun short(): Short = decodeWithComputed { ShortDecoder(endianness) }
+
+    @JvmName("decodeByte")
+    override fun int(): Int = decodeWithComputed { IntDecoder(endianness) }
+
+    @JvmName("decodeByte")
+    override fun long(): Long = decodeWithComputed { LongDecoder(endianness) }
+
+    @JvmName("decodeByte")
+    override fun float(): Float = decodeWithComputed { FloatDecoder(endianness) }
+
+    @JvmName("decodeByte")
+    override fun double(): Double = decodeWithComputed { DoubleDecoder(endianness) }
+
+    @JvmName("decodeByte")
+    override fun boolean(): Boolean = decodeWithComputed { BooleanDecoder() }
+
+    @JvmName("decodeString")
+    override fun string(): String = decodeWithComputed {
+        throw AssertionError("A String decoder should always be present")
+    }
+
+    @JvmName("decodeSelfOrNull")
+    override fun selfOrNull(): E? = selfOrNull(computed { BooleanDecoder() })
+
+    override fun <R> selfCollectionOrNull(collector: Collector<E, *, R>): R? =
+        selfCollectionOrNull(collector, computed { IntDecoder(endianness) }, computed { BooleanDecoder() })
+
+    override fun <R> selfCollectionOrNull(collector: Collector<E, *, R>, sizeDecoder: Decoder<Int>): R? =
+        selfCollectionOrNull(collector, sizeDecoder, computed { BooleanDecoder() })
+
+    @JvmName("decodeSelfCollection")
+    override fun <R> selfCollection(collector: Collector<E, *, R>): R =
+        selfCollection(collector, computed { IntDecoder(endianness) })
+
+    @Suppress("UNCHECKED_CAST")
+    private inline fun <reified T> decodeWithComputed(noinline block: () -> Decoder<T>): T {
+        val decoder = cache.computeIfAbsent(T::class.java) { block() } as Decoder<T>
+        return decode(decoder)
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private inline fun <reified T> computed(noinline block: () -> Decoder<T>): Decoder<T> =
+        cache.computeIfAbsent(T::class.java) { block() } as Decoder<T>
+
+    private inner class SelfDecoder : Decoder<E> {
+
+        override fun decode(input: DecoderInput): Decoder.State<E> = if (accumulator.hasRecursionElement()) {
+            val element = accumulator.getFromRecursion()
+            accumulator.add(element)
+            currentIndex++
+            Decoder.State.Done(element.tryCast<E>())
+        } else {
+            throw RecursionMarker
+        }
+
+        override fun reset() = Unit
+
+    }
+
+    internal fun reset() {
+        self.reset()
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private fun <T> Any?.tryCast(): T = this as? T
+        ?: throw IllegalStateException(
+            """
+        Error while trying to cast $this.
+        This error may have been caused because different calls have been made in the the scope for the same object,
+        between two calls of the block. Calls in the scope must be consistent for the same object between two calls.
+            """.trimIndent()
+        )
```

Comment:
```suggestion
    @Suppress("UNCHECKED_CAST")
    private fun <T> Any?.tryCast(): T = this as? T
        ?: throw IllegalStateException("""
            Error while trying to cast $this.
            This error may have been caused because different calls have been made in the the scope for the same object,
            between two calls of the block. Calls in the scope must be consistent for the same object between two calls.
        """.trimIndent())
```